Add vector space representation for magnitudes

chiphogg opened the pull request:
This is the beginning of the implementation for #300, laying the first foundations.  This seems like a good point to check in and align on direction before we get too far along.  For general context:

_Things we include here:_
- Concepts for `Magnitude`, and `BasePower` (a Magnitude is made up of a canonicalized parameter pack of BasePowers).
- End user interfaces for making magnitudes:
  - `as_magnitude<ratio>()`: the main way to make a Magnitude from a rational number.  This handles the "heavy lifting" of converting to a prime factorization.
  - `base_power<T>{ratio}`: we handle irrational bases by introducing a type `T`, with a static `long double` member `value` which must be positive.
- Enhanced `ratio` functionality: addition, negation, and implicit construction.

_Work yet undone:_
- A mechanism for _applying_ a magnitude to a given value.  (We'll probably want to break out the rational and irrational parts separately, so we can give exact results as often as possible.  We'll also want to think carefully about what types we use in intermediate computations---for example, not everyone will want to use `long double`, and even `double` can be problematic for some embedded applications.)
- Actually migrating existing uses to use Magnitude instead of ratio.

Once we get to something that looks like the right direction, I can flesh out these other items.

chiphogg commented on src/core/include/units/magnitude.h:
```diff
@@ -0,0 +1,331 @@
+// The MIT License (MIT)
+//
+// Copyright (c) 2018 Mateusz Pusz
```
Is this right?  I note that it is not 2018, and I am not Mateusz Pusz, but I tried following the style in the other files.

chiphogg replied:
Just checking in again about this copyright notice question.

mpusz replied:
I just copy the same to all the files and I noticed that other contributors do the same.

I do not mind changing this approach but frankly speaking, I do not have a good idea how to do it. All of the contributors are enumerated automatically by GitHub already. I also listed the most active contributors in the citation file (at least 50 commits).

Please let me know in case you have some suggestions.

JohelEGP commented:
The code looks good. I'd appreciate an example that shows what can be done with that has been implemented so far.

chiphogg commented:
> The code looks good. I'd appreciate an example that shows what can be done with that has been implemented so far.

With what's been implemented so far?  The answer is "not much", I'm afraid.  :slightly_smiling_face:

The unit tests give some examples.  You can make Magnitudes via `make_ratio<...>()` and `make_base_power<...>()`, multiply and divide them, and compare them for equality.  But as for more useful operations, here's where I see this going.

**The point of a Magnitude is to perform unit conversions.**  In terms of bits and bytes, we'll convert a value in one Rep to a value in another (possibly different) Rep.  What properties should this conversion have?  Here are some (and I doubt this is an exhaustive list):

- It should be fast: we should do as much as we can at compile time.
- It should be accurate: if an exact answer is computable, we should compute it.
  - Here's a specific example to show what I have in mind.  Imagine converting a floating point value from inches to feet.  This means dividing our value by 12.  Every value which is an integer multiple of 3 has an _exact_ answer, and we should always produce this.  (So we must _divide by 12_, instead of multiplying by the inexact floating point representation of 1/12.)
- It should be thoughtful about intermediate types.  Generally, we should perform our computations in the widest possible type, given the input and output types.  But in a nod to our embedded customers, who may not have hardware acceleration for, say, `double`, we'd like to keep `float` computations fully within `float` types.

What does a solution look like?  Probably something like this:

```cpp
template<typename NewRep, typename OldRep, Magnitude M>
struct Converter;
// Specializations provide:
// static constexpr NewRep convert(OldRep value);

template<typename NewRep, typename OldRep>
constexpr NewRep convert(OldRep value, Magnitude auto factor) {
  return Converter<NewRep, OldRep, decltype(factor)>::convert(value);
}

// Example usage callsite:
constexpr auto length_ft = convert<double>(9 /* inches */, make_ratio<1, 12>());
// should produce `double{0.75}`.
```

Granted, this computation uses raw arithmetic types, which looks weird in a units library context.  But in the real implementation, all of these uses would be encapsulated inside of `quantity` types and their API functions.

As an alternative, I considered simply implementing multiplication with arbitrary Magnitudes and arithmetic types (the way a Magnitude can multiply another Magnitude), but this will not be sufficient because we need to customize the output type.

I wish I had more usefulness out of the gate... but, this seemed like the right size for a "first chunk" that could help us get aligned on where we're headed, before investing more significant effort.  (Not to mention, navigating the above policies in the first PR felt like way too much cognitive load!  :sweat_smile:)

JohelEGP commented:
> > The code looks good. I'd appreciate an example that shows what can be done with that has been implemented so far.
> 
> With what's been implemented so far? The answer is "not much", I'm afraid. slightly_smiling_face

> I wish I had more usefulness out of the gate... but, this seemed like the right size for a "first chunk" that could help us get aligned on where we're headed, before investing more significant effort. (Not to mention, navigating the above policies in the first PR felt like way too much cognitive load! sweat_smile)

That makes sense. No examples because nothing can be done yet. On retrospective, it's obvious the feature hasn't been integrated into `mp-units`' framework.

> * It should be thoughtful about intermediate types.  Generally, we should perform our computations in the widest possible type, given the input and output types.  But in a nod to our embedded customers, who may not have hardware acceleration for, say, `double`, we'd like to keep `float` computations fully within `float` types.

That's how it currently works in practice. Only integers are converted by going through their common type with `std::intmax_t`, which is a noop for floating point types.

> _Work yet undone:_
> 
> * A mechanism for _applying_ a magnitude to a given value.  (We'll probably want to break out the rational and irrational parts separately, so we can give exact results as often as possible.  We'll also want to think carefully about what types we use in intermediate computations---for example, not everyone will want to use `long double`, and even `double` can be problematic for some embedded applications.)
> 
> * Actually migrating existing uses to use Magnitude instead of ratio.

Also related: #195. I suppose the most obvious thing to do would be generalizing `ratio`, as has been pointed out.

chiphogg commented:
The build failed on MSVC 14.2.  I don't know how to run a given build plan locally, but I pushed up a new commit that _should_ fix it.  I'll still need a maintainer to approve re-running the checks, though.

JohelEGP commented on src/core/include/units/magnitude.h:
```diff
@@ -261,10 +261,11 @@ struct product<magnitude<base_power<Base, Pow1>, Tail1...>,
                magnitude<base_power<Base, Pow2>, Tail2...>>
 {
   using tail_product = product_t<magnitude<Tail1...>, magnitude<Tail2...>>;
+  static inline constexpr auto Pow = Pow1 + Pow2;
```
This `inline` is redundant.

chiphogg replied:
Fixed, here and elsewhere.

mpusz commented on test/unit_test/runtime/magnitude_test.cpp:
```diff
@@ -0,0 +1,403 @@
+  SECTION("Products handle pi correctly")
+  {
+    CHECK(
+        make_base_power<pi>() * make_ratio<2, 3>() * make_base_power<pi, -1, 2>() ==
+        magnitude<int_base_power<2>, int_base_power<3, -1>, base_power<pi, ratio{1, 2}>>{});
```
Does it mean that the `Den` part of `ratio` is only needed for floating-point bases (for integrals you can always extract num and den to separate magnitudes)? Also, it seems that `exp` of `units::ratio` is never used here. Maybe `units::ratio` is not the best to implement this (for sure it makes the type name longer in the compilation errors).

mpusz replied:
Also, the fact of having `exp` in `units::ratio` complicates many operations which could be faster to compile (or more accurate) if it was not there. If we can simplify things there it would be great.

mpusz replied:
> `Den` is not 1, and there's no other way to represent this, right?

Right, I forgot about `sqrt`...

mpusz replied:
I would love to be able to get rid of `exp` from `ratio` if only possible.

chiphogg replied:
I don't know if it will work, because I haven't looked at all the use cases in the codebase.  But I plan to _try_ to get rid of it, and I expect to succeed!

mpusz commented:
Nice work @chiphogg 👍🏻

mpusz commented:
> The build failed on MSVC 14.2. I don't know how to run a given build plan locally, but I pushed up a new commit that _should_ fix it. I'll still need a maintainer to approve re-running the checks, though.

Unfortunately, this is how it works for new contributors. After some more commits the CI will start automatically.

JohelEGP commented on src/core/include/units/magnitude.h:
```diff
@@ -0,0 +1,337 @@
+
+// Default implementation: most things are not magnitudes.
+template<typename T>
+struct is_magnitude: std::false_type {};
```
It should be the same.

chiphogg replied:
Done.

chiphogg commented:
> > The build failed on MSVC 14.2. I don't know how to run a given build plan locally, but I pushed up a new commit that _should_ fix it. I'll still need a maintainer to approve re-running the checks, though.
> 
> Unfortunately, this is how it works for new contributors. After some more commits the CI will start automatically.

No problem.  This makes me glad I'm breaking out the changes into multiple PRs.  :slightly_smiling_face: 

Do you have any advice for me to test my changes locally?  I can read the errors, but I'm basically just guessing as to whether my changes fix them.  (I pushed some commits that _should_ get the build to green, but those builds need to be enabled.  I'll look through the other comments in detail when I get a chance.)

JohelEGP commented:
Maybe you can get the workflows to run on your fork.

chiphogg commented:
> Maybe you can get the workflows to run on your fork.

Oh, I just had an idea---maybe we can turn this PR into only the `ratio` changes (addition and negation)?  That should be easier to get through code review.  Then I could make another PR with the updated `magnitude` changes after the stripped down version of this lands, and then I wouldn't be blocked on running the builds.

The downside is that we would end up losing the review comments on magnitude.  But I can still respond to them one round before we make this change, and then the first iteration of the new/"real" magnitude PR would take them into account.

@mpusz, what do you think of this idea?

chiphogg commented:
> > Maybe you can get the workflows to run on your fork.
> 
> Oh, I just had an idea---maybe we can turn this PR into only the `ratio` changes (addition and negation)? That should be easier to get through code review. Then I could make another PR with the updated `magnitude` changes after the stripped down version of this lands, and then I wouldn't be blocked on running the builds.
> 
> The downside is that we would end up losing the review comments on magnitude. But I can still respond to them one round before we make this change, and then the first iteration of the new/"real" magnitude PR would take them into account.
> 
> @mpusz, what do you think of this idea?

Actually---here's a better plan.

- We work through all the comments on this PR, until we get to where we're happy.
- Use local builds only for the testing: don't worry about MSVC failures I can't reproduce.
- If the builds pass when we get there, great!  We can land this.  Otherwise, we can consider falling back to make this a simpler `ratio` PR, and pushing the `magnitude` stuff off to a future PR, where I can iterate on all the builds without being blocked on another human.

So, for clarity: I plan to work through all the existing review comments and update this PR/respond to comments when it's ready.

mpusz commented:
> No problem. This makes me glad I'm breaking out the changes into multiple PRs. 🙂

I do not think this makes any change here. The first 3-4 pushes from a new contributor have to be accepted for the CI by the maintainer so some anonymous users are not able to kill GitHub CI system with a flood of commits. I think that we reached this limit already or will reach it soon so after that everything will be triggered automatically.

mpusz commented:
> Do you have any advice for me to test my changes locally? I can read the errors, but I'm basically just guessing as to whether my changes fix them. (I pushed some commits that should get the build to green, but those builds need to be enabled. I'll look through the other comments in detail when I get a chance.)

I find Windows with WSL the most robust for OSS projects as it allows me to test gcc, clang, and MSVC on the same repo on the same machine without the need to switch OSes.

mpusz commented:
> Oh, I just had an idea---maybe we can turn this PR into only the ratio changes (addition and negation)? That should be easier to get through code review.

I do not think it is needed. I reviewed the code and it looks really OK 👍🏻 
I just had some comments about the compile-time performance and some `is_*` functions that seem to be used only for unit tests.

chiphogg commented:
I'm still working on addressing the review comments---it's a big change, but a fun one.  Exhilarating!

While I'm working: @mpusz, how do you feel about marking the `ratio` constructor as `explicit(false)`?  I can't really see the harm in passing an integral value to an API expecting a `ratio`.  What's more, it seems nice to be able to compare, say, a `ratio` for equality with an `int`.

My motivation for asking is that it makes some syntax much less clunky.  Here's an example for the prime factorization of 792:

```cpp
// With implicit ratio constructor:
magnitude<base_power{2, 3}, base_power{3, 2}, base_power{11}>{}

// Without implicit ratio constructor:
magnitude<base_power{2, ratio{3}}, base_power{3, ratio{2}}, base_power{11}>{}
```

I can get by without it, but I think it's nicer as an implicit constructor (unless somebody raises an objection).

chiphogg commented on src/core/include/units/magnitude.h:
```diff
@@ -0,0 +1,337 @@
+ * all exponents are zero.  To achieve this, we use arbitrarily many prime numbers as basis vectors, and also various
+ * irrational numbers such as pi.
+ *
+ * @tparam Base  A type representing the basis vector.  Must have a numeric static `value` member.
```
I got this working for the prime bases (and it's really nice!), but I'm running into difficulties with floating point.  It looks like `g++-10` won't accept template parameters of floating point type:

```
/home/chogg/github/units/test/unit_test/runtime/magnitude_test.cpp:97:17: error: ‘double’ is not a valid type for a template non-type parameter
   97 | template<double x>
      |                 ^
/home/chogg/github/units/test/unit_test/runtime/magnitude_test.cpp:98:18: error: expected type-specifier before numeric constant
   98 | using double_v = 2 * x;
```

Is this right?  Do we use floating point template parameters anywhere else in the library?

I'm going to try to work around this when I pick it up next time.

chiphogg replied:
Got it!  I'm polishing for review still, with a lot of work left to do, but I'm over the hump.

Re-implementing the product using `if constexpr` was mind-expanding!

JohelEGP replied:
See https://gcc.gnu.org/projects/cxx-status.html:
```
10 (no floating point, union, or subobject template args)
11 (complete)
```

chiphogg replied:
The current approach is to use a _member_ for the _prime_ bases, but a _type_ for the _irrational_ bases.  We access them in a uniform way via the `.get_base()` function (which could be made a free function if we prefer).

One benefit of using a _type_ for the irrational bases is that it makes it clear that we are doing _symbolic_ reasoning.  It gives us a little distance from the value we're storing, which is only ever an approximation anyway.  With a value, if we ever changed to a higher precision storage, we'd have to update a ton of callsites.  With a type, we're much better protected.

mpusz replied:
Well, we can always create an alias in one place and use it as a type of value in all the implementation. This should improve maintenance in the future.

chiphogg reviewed (commented):
The work is still in progress (no update yet), but I wanted to check in about floating point NTTPs.

It looks like they don't work in g++10.  If that's right, then we'll have to use a _type_ for the base of `base_power`, rather than an NTTP.  Does that seem right to you?

chiphogg commented on src/core/include/units/magnitude.h:
```diff
@@ -0,0 +1,337 @@
+ * @tparam Power  The rational power to which the base is raised.
+ */
+template<typename Base, ratio Power = ratio{1}>
+struct base_power {
```
Done.  I hadn't fully realized the possibilities which C++20 brought---this is so much cleaner!

chiphogg commented on src/core/include/units/magnitude.h:
```diff
@@ -0,0 +1,337 @@
+concept BasePower = is_base_power_v<T>;
+
+/**
+ * @brief  A representation for positive real numbers which optimizes taking products and rational powers.
```
The new version checks the individual base powers for validity.  Since valid base powers always have a positive base, yes, we do now check this with a constraint.

chiphogg commented on src/core/include/units/magnitude.h:
```diff
@@ -0,0 +1,337 @@
+template<Magnitude M>
+struct inverse;
+template<Magnitude M>
+using inverse_t = typename inverse<M>::type;
```
Done.  (In this case, I made an `inverse(Magnitude auto)` function.  Later on, we'll probably change to `pow<ratio>(Magnitude auto)`.)

chiphogg commented on src/core/include/units/magnitude.h:
```diff
@@ -0,0 +1,337 @@
+template<Magnitude... Mags>
+struct product;
+template<Magnitude... Mags>
+using product_t = typename product<Mags...>::type;
```
I had thought I would need these "old school" metaprogramming constructs in order to deduce the types.  I was wrong!  The new version uses functions with `if constexpr` branches.  We still need a few special case overloads, but overall this is much cleaner.

chiphogg commented on src/core/include/units/magnitude.h:
```diff
@@ -0,0 +1,337 @@
+ * @brief  Compute the quotient of 2 Magnitudes.
+ */
+template<Magnitude T, Magnitude U>
+using quotient_t = product_t<T, inverse_t<U>>;
```
Done.

chiphogg commented on src/core/include/units/magnitude.h:
```diff
@@ -0,0 +1,337 @@
+ */
+template<std::intmax_t N>
+struct prime_base : std::integral_constant<std::intmax_t, N> {
+  static_assert(detail::is_prime(N));
```
Done.  (Well, in spirit, anyway---based on other feedback, I got rid of `prime_base` in favor of NTTP values.)

chiphogg commented on src/core/include/units/magnitude.h:
```diff
@@ -0,0 +1,337 @@
+
+// Default implementation: most things are not base powers.
+template<typename T>
+struct is_base_power: std::false_type {};
```
I couldn't figure out how to do an `if constexpr`-type implementation, but I did convert these to variable templates with specializations.

chiphogg commented on src/core/include/units/magnitude.h:
```diff
@@ -0,0 +1,337 @@
+
+// To be a valid base power, one must be a base_power<B, E>, where B has a static value member which is positive.
+template<typename B, ratio E>
+  requires requires() { B::value; }
```
The current implementation constrains the base to be either `int` or (indirectly) `long double`.

chiphogg commented on src/core/include/units/magnitude.h:
```diff
@@ -0,0 +1,337 @@
+  }(sizeof...(Ts));
+
+  // Compare zero or more pairs of neighbours as needed.
+  return [&ts, &pred]<std::size_t... Is>(std::integer_sequence<std::size_t, Is...>) {
```
Done.

chiphogg commented on src/core/include/units/magnitude.h:
```diff
@@ -0,0 +1,337 @@
+// Specialization for the prime factorization of larger numbers (recursive case).
+template<std::intmax_t N>
+struct prime_factorization {
+  static_assert(N > 1, "Can only factor positive integers.");
```
Done.

chiphogg commented on test/unit_test/runtime/magnitude_test.cpp:
```diff
@@ -0,0 +1,403 @@
+        magnitude<int_base_power<3, -4>, int_base_power<5>>>);
+  }
+}
+
```
I moved all of the `is_...` to the `detail` namespace.  There are fewer of them now, I think, and they are only used as implementation details (mostly for concepts).

chiphogg commented on test/unit_test/runtime/magnitude_test.cpp:
```diff
@@ -0,0 +1,403 @@
+{
+  SECTION("Equivalent to std::integral_constant for integer bases")
+  {
+    CHECK(make_base_power<prime_base<2>>() == make_ratio<2>());
```
I'm happy to do this if you want, but I like how the hierarchical setup of `TEST_CASE` and `SECTION` comments communicates semantically to the reader.  I'm hesitant to throw that information away.

The main benefit of making these static asserts is to verify that we can perform these operations at compile time (e.g., didn't forget a `constexpr` somewhere), right?

I feel like what we really want is a sort of `CHECK_STATIC(...)`, which performs the usual `CHECK`, but also a `static_assert` under the hood.  Do you know if there's anything like that?

mpusz replied:
> The main benefit of making these static asserts is to verify that we can perform these operations at compile time (e.g., didn't forget a constexpr somewhere), right?

In general yes. However, another nice property of the project is that if it compiles fine it is correct (aside from text formatting that has to be verified at runtime). No unit tests have to be run to check that.

chiphogg reviewed (commented):
I think I've addressed most of the feedback---at least, enough for the next round of review.  Let me know how it's shaping up!

JohelEGP commented on test/unit_test/runtime/magnitude_test.cpp:
```diff
@@ -385,27 +262,79 @@ TEST_CASE("is_prime detects primes")
+  }
+}
+
 TEST_CASE("pairwise_all evaluates all pairs")
```
@mpusz Have you thought about white-box testing in a modules world? I personally use a `.details` module which gets the same treatment as public API modules in terms of tests.

mpusz replied:
Actually, I do not have much experience with C++20 modules so far. If you have any then I will be glad to learn and benefit from your suggestions. The only action point I thought of so far is to try to move all `namespace detail` implementation to separate header files in the `bits` directory so it will be easier to define exported and private module interfaces.

JohelEGP replied:
I move things to a common `.details` module only when they're used across different modules. For something like `pairwise_all`, I choose to leave it in the only module it is used and test it indirectly through the public API. In this case, besides testing `magnitude` (the to-be replacement of `ratio`) normally, I'd add additional test cases, if necessary, to prove that its use of `pairwise_all` works as expected.

mpusz replied:
Does it mean that you use modules already? I thought that they are ready only on MSVC and if I remember correctly you do not use this compiler. Also, there is still no support for them in CMake.

If modules are already "usable" on major compilers we should add support for them and check how it addresses compile-time issues for the SI definitions.

JohelEGP replied:
I use https://github.com/JohelEGP/jegp.cmake_modules/#project-modules to some amount of success. There are many bugs. GCC is particularly plagued with bugs relating to the C++ Standard Library. Can't wait to have `std` as a module.

JohelEGP commented on test/unit_test/runtime/magnitude_test.cpp:
```diff
@@ -385,27 +262,79 @@ TEST_CASE("is_prime detects primes")
+    CHECK(pairwise_all{std::less{}}(1, 1.5));
+    CHECK(pairwise_all{std::less{}}(1, 1.5, 2));
+    CHECK(!pairwise_all{std::less{}}(1, 2.0, 2));
+    CHECK(!pairwise_all{std::less{}}(1, 2.5, 2));
```
This would fail for `-1U, 1`. Should it use `cmp_less` from `hacks.h`?

chiphogg replied:
It never would have occurred to me to write something like `-1U`.  Based on how literals work, I expect this to be "the negation of `1U`".  So it should be a pretty large positive number.

So: what would it mean to "fail"?

I would expect the result from the function to be `false`, because we're comparing a large positive number to `1`.  But what result would you expect for `pairwise_all{std::less{}}(-1U, 1)`, and why?

---

To answer your other question: I'm only too happy to use `cmp_less` in the implementation.  I don't think it matters much either way for this use case, but maybe there's another use case I haven't yet thought of.

JohelEGP replied:
Sorry, that should have been `-1, 1U`. `-1 < 1U` is `false`.

chiphogg replied:
That makes much more sense, thanks!

I'll add some test cases for `strictly_increasing` to cover this use case when I get back to my work computer.

mpusz commented:
> how do you feel about marking the ratio constructor as explicit(false)?

I generally like the idea. I have only one doubt. Maybe it would be better to allow `base_power` to take either an `int` or `ratio` as a second template parameter (with constrained `auto` parameter). With that, a template type printed in many error messages and in a debugger will be much shorter, right? Or most of the types will end up with a `ratio` anyway?

mpusz commented:
> If that's right, then we'll have to use a type for the base of base_power, rather than an NTTP. Does that seem right to you?

It works for gcc-10. The question is would it be a big problem to have both implementations (a type only for gcc-10)? I am asking because having it used and tested would simplify a standardization effort later on...

mpusz commented on src/core/include/units/magnitude.h:
```diff
@@ -0,0 +1,360 @@
+ * @brief  A simple way to create a Magnitude representing a rational power of pi.
+ */
+template<ratio Power>
+constexpr auto pi_to_the() { return magnitude<base_power<pi_base>{Power}>{}; }
```
I am not sure if we need that. This will not be used too often in the library (only for angles). Maybe moved to unit tests file if needed.

mpusz commented on src/core/include/units/magnitude.h:
```diff
@@ -0,0 +1,360 @@
+// Magnitude product implementation.
+
+// Base cases, for when either (or both) inputs are the identity.
+constexpr auto operator*(magnitude<>, magnitude<>) { return magnitude<>{}; }
+constexpr auto operator*(magnitude<>, Magnitude auto m) { return m; }
+constexpr auto operator*(Magnitude auto m, magnitude<>) { return m; }
+
+// Recursive case for the product of any two non-identity Magnitudes.
+template<BasePower auto H1, BasePower auto... T1, BasePower auto H2, BasePower auto... T2>
+constexpr auto operator*(magnitude<H1, T1...>, magnitude<H2, T2...>) {
```
Operators could be implemented as hidden friends.

chiphogg replied:
Good point!  I typically prefer hidden friends, but for some reason it hadn't occurred to me here.  I started working on this, but it seemed like I'd need multiple specializations of `magnitude` for different patterns of parameter packs.  If that's true, I'm not sure it's worth the added complexity.  But maybe there's a strategy for this which I didn't see.

mpusz replied:
Let's merge it as it is and then possibly try to refactor.

mpusz commented on src/core/include/units/magnitude.h:
```diff
@@ -0,0 +1,360 @@
+  else { return bp.get_base() > 0; }
+}
+
+// A function object to apply a predicate to all consecutive pairs of values in a sequence.
+template<typename Predicate>
+struct pairwise_all {
+  Predicate predicate;
+
+  template<typename... Ts>
+  constexpr bool operator()(Ts&&... ts) const {
+    // Carefully handle different sizes, avoiding unsigned integer underflow.
+    constexpr auto num_comparisons = [](auto num_elements) {
+      return (num_elements > 1) ? (num_elements - 1) : 0;
+    }(sizeof...(Ts));
+
+    // Compare zero or more pairs of neighbours as needed.
+    return [this]<std::size_t... Is>(std::tuple<Ts...> &&t, std::index_sequence<Is...>) {
+      return (predicate(std::get<Is>(t), std::get<Is + 1>(t)) && ...);
+    }(std::make_tuple(std::forward<Ts>(ts)...), std::make_index_sequence<num_comparisons>());
+  }
+};
+
+// Deduction guide: permit constructions such as `pairwise_all{std::less{}}`.
+template<typename T>
+pairwise_all(T) -> pairwise_all<T>;
```
Well, all of the C++ standard algorithms take the predicate as a function argument rather than a template parameter 😉 so I think it is less surprising to the user than introducing a stateful functor. Also, one function template should be faster to compile than two (class and member function). Although the class template for `std::less{}` will probably be memoized in many cases.

chiphogg replied:
Good points.  Although I think the C++ standard algorithms take iterators (or possibly ranges), as opposed to a variadic number of arguments.  I think including the predicate along with a variadic number of arguments feels different, and the separation feels more important.

Regardless, I think this is something worth revisiting in its own PR later if we choose.

mpusz reviewed (approved):
Thanks a lot @chiphogg!!!

chiphogg commented:
My pleasure.  And we got rid of that pesky (spurious IMO) MSVC 14.2 compiler error, too!

So... how do we actually land this?  I don't think I have access to do that.

chiphogg commented:
By the way, based on unresolved comments visible from the main page, I've collected this list of follow-up items:

- [ ] Move tests to compile time
- [ ] Work to remove `exp` from `ratio`
- [ ] Consider supporting plain `int` base powers
- [ ] Consider refactoring `operator*` et al to be hidden friends

Let me know if I've missed any.
